### PR TITLE
#22996 initial changes

### PR DESCRIPTION
**Pull request by dsilvam**
Adds the variantName to the ID of the ES document. 
Makes "variant" field a keyword so it doesn't get tokenized. 
Integration tests.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/common/reindex/BulkProcessorListener.java [hunk 1]
```diff
@@ -95,6 +103,18 @@ public void afterBulk(final long executionId, final BulkRequest request, final B
         }
     }
 
+    static String getMatchingReservedIdIfAny(String id) {
+        String matchingReservedId = null;
+
+        for (String reservedId : RESERVED_IDS) {
+            if(id.contains(reservedId)) {
+                matchingReservedId = reservedId;
```
Comment (victoralfaro-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/common/reindex/BulkProcessorListener.java:111]: Shouldn't we return/break the loop execution?

Comment (dsilvam) [reply dotCMS/src/main/java/com/dotmarketing/common/reindex/BulkProcessorListener.java:111]: Good catch! done

**Review by victoralfaro-dotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java [hunk 2]
```diff
@@ -872,15 +874,19 @@ public void removeContentFromIndex(final Contentlet content) throws DotHibernate
...
         final List<Language> languages = APILocator.getLanguageAPI().getLanguages();
+        final List<Variant> variants = APILocator.getVariantAPI().getVariants();
         final IndiciesInfo info = APILocator.getIndiciesAPI().loadIndicies();
 
         // delete for every language and in every index
         for (Language language : languages) {
             for (final String index : info.asMap().values()) {
-                final String id = entry.getIdentToIndex() + "_" + language.getId();
+                for(final Variant variant: variants) {
+                    final String id = entry.getIdentToIndex() + "_" + language.getId()
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java:884]: Use StringPool.UNDERLINE

File: dotCMS/src/main/java/com/dotcms/variant/VariantAPIImpl.java [hunk 3]
```diff
@@ -114,4 +115,9 @@ public Optional<Variant> get(final String name) throws DotDataException {
 
         return variantFactory.get(name);
     }
+
+    @Override
+    public List<Variant> getVariants() throws DotDataException {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/variant/VariantAPIImpl.java:120]: add close db if opened

File: dotCMS/src/main/java/com/dotcms/variant/VariantFactoryImpl.java [hunk 4]
```diff
@@ -119,4 +122,11 @@ public Optional<Variant> get(final String name) throws DotDataException {
             return variantFromDataBase;
         }
     }
+
+    @Override
+    public List<Variant> getVariants() throws DotDataException {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/variant/VariantFactoryImpl.java:127]: Cache?

File: dotCMS/src/main/java/com/dotmarketing/common/reindex/BulkProcessorListener.java [hunk 5]
```diff
@@ -95,6 +103,19 @@ public void afterBulk(final long executionId, final BulkRequest request, final B
         }
     }
 
+    static String getMatchingReservedIdIfAny(String id) {
+        String matchingReservedId = null;
+
+        for (String reservedId : RESERVED_IDS) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/common/reindex/BulkProcessorListener.java:109]: set to final

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java (hunk 2 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ContentletIndexAPIImpl.java:884]: Done

File: dotCMS/src/main/java/com/dotcms/variant/VariantAPIImpl.java (hunk 3 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/variant/VariantAPIImpl.java:120]: Done

File: dotCMS/src/main/java/com/dotmarketing/common/reindex/BulkProcessorListener.java (hunk 5 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotmarketing/common/reindex/BulkProcessorListener.java:109]: Done

**Review by freddyucv — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/variant/VariantFactoryImpl.java [hunk 6]
```diff
@@ -119,4 +122,11 @@ public Optional<Variant> get(final String name) throws DotDataException {
...
+        return TransformerLocator.createVariantTransformer(new DotConnect()
+                .setSQL(VARIANT_SELECT_ALL_QUERY)
```
[the remark this answers is not in the document]
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/variant/VariantFactoryImpl.java:129]: Done

**Pull request closed without merging**